### PR TITLE
test: Add tests for is_in_node_range

### PR DESCRIPTION
**Pull request by last-partizan**
~~This fixes no indents at the end of python files~~

~~(Bonus, and passes 18 tests (+2))~~

This adds tests for is_in_node_range

Refs #1136

### CODE REVIEW

**Review by kyazdani42 — APPROVED** [review]
looks good to me, @vigoux do you remember why the base implementation is so complex ?

**Comment by theHamsta** [comment]
Because actually is_in_node_range should not include end row/col because Tree-Sitter are end exclusive

**Review by theHamsta — CHANGES_REQUESTED** [review]
Tree-Sitter ranges are end exclusive

**Comment by last-partizan** [comment]
@theHamsta are you sure about that?

image

Looking at this, `[1, 8]` is last character of function definition, body block and pass statement. Why it should be excluded?

Is there examples where this could break something? for python it fixes #1136, and i haven't noticed anything suspicious using it today. But, maybe there was something suspicious, and i discarded it thinking it was broken before, i'll try to remember what it was and test if reverting fixes it.

**Comment by last-partizan** [comment]
And, if ranges are end exclusive, then root cause of invalid indents at the end of python files - is that we cannot find a root.

I just checked https://tree-sitter.github.io/tree-sitter/playground, and it has top-level root `module`, and neovim does not, maybe this is causing bug?

image

**Comment by last-partizan** [comment]
Turns out i missed important point, this is not simple `check point is belongs to rectangle`, with lines and cols we need to verify all those edge cases.

While I was investigating, I added some tests. And i think they should be useful, but that's my first time writing tests in lua, and there was no good example on how to write unit-test. Please, review this draft version and i'll improve them with feedback.

File: tests/unit/ts_utils_spec.lua [hunk 1]
```diff
@@ -0,0 +1,34 @@
+local tsutils = require "nvim-treesitter.ts_utils"
+
+describe("Test is_in_node_range", function()
```
Comment (vigoux) [inline tests/unit/ts_utils_spec.lua:3]: This should be `is_node_in_range` alone.

File: tests/unit/ts_utils_spec.lua [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    {{0, 3, 0, 5}, 0, 1, false},
+    {{0, 3, 0, 5}, 0, 3, true},
+    {{0, 3, 0, 5}, 0, 5, false},
+  }
+  for _, param in ipairs(params) do
+    local r, line, col, expected = unpack(param)
+    it(string.format("%s, %s, %s -> %s", table_repr(r), line, col, expected), function ()
+      is_in_node_range(unpack(param))
+    end)
+  end
```
Comment (vigoux) [inline tests/unit/ts_utils_spec.lua:33]: What we (at nvim, but not here yet), tend to do, is wrap those in `it` blocks (same signature as `describe`) with the "semantic" of this test, to help when the test fails afterwards, do you think you can make that a thin?

**Review by vigoux — COMMENTED** [review]
Pretty good for the test addition, we should enhance that for future work and add more tests as we go, following this example.

File: tests/unit/ts_utils_spec.lua (hunk 2 above)
Comment (last-partizan) [reply tests/unit/ts_utils_spec.lua:33]: I don't understand, could you give some examples?

Comment (vigoux) [reply tests/unit/ts_utils_spec.lua:33]: Maybe do something like : 
```
it('correctly gets nodes on the edge', function()
  -- Put your test here.
end)
```

Comment (last-partizan) [reply tests/unit/ts_utils_spec.lua:33]: Oh, describe each of the calls instead of just plain params list. Yes, i'll do that.

Comment (last-partizan) [reply tests/unit/ts_utils_spec.lua]: Fixed :)

I actually like it this way :)

btw, is there any docs on tree-sitter node range end-exclusivity? so i can link that in test description.

File: tests/unit/ts_utils_spec.lua [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  function test_is_in_node_range(line, col, expected)
+    local node = {
+      range = function ()
+        return unpack({0, 3, 2, 5})
+      end
+    }
+    return tsutils.is_in_node_range(node, line, col)
+  end
+
+  it("false before node start", function ()
```
Comment (vigoux) [inline tests/unit/ts_utils_spec.lua:14]: Maybe add a `returns` here, this is just wording but it really sounds better afterwards.

**Review by vigoux — COMMENTED** [review]
We are getting there, just a final comment and we'll be good to go.

File: tests/unit/ts_utils_spec.lua (hunk 3 above)
Comment (last-partizan) [reply tests/unit/ts_utils_spec.lua:14]: Yeah, should i add it for each `it`, or can i do this just for `describe`?

Comment (vigoux) [reply tests/unit/ts_utils_spec.lua]: I would add it to each `it`, so that it reads as `it('returns something',...)`.

**Review by vigoux — APPROVED** [review]
A-we-som

**Review by theHamsta — APPROVED** [review]
